src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/13.diff.sql

```diff
@@ -0,0 +1,424 @@
+    @startResourceSurrogateId bigint,
+    @endResourceSurrogateId bigint,
+    @batchSize int
+AS
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    DELETE Top(@batchSize) FROM dbo.ResourceWriteClaim
+    WHERE ResourceSurrogateId >= @startResourceSurrogateId AND ResourceSurrogateId < @endResourceSurrogateId
+
+    COMMIT TRANSACTION
+
+    return @@rowcount
+GO
+
+
+/*************************************************************
+    Stored procedures for batch delete ResourceParams
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     BatchDeleteResourceParams
+--
+-- DESCRIPTION
+--     Batch delete ResourceParams
+--
+-- PARAMETERS
+--     @tableName
+--         * Resource params table name
+--     @resourceTypeId
+--         * Resource type id
+--     @startResourceSurrogateId
+--         * The start ResourceSurrogateId
+--     @endResourceSurrogateId
+--         * The end ResourceSurrogateId
+--     @batchSize
+--         * Max batch size for delete operation
+CREATE OR ALTER PROCEDURE dbo.BatchDeleteResourceParams
+    @tableName nvarchar(128),
+    @resourceTypeId smallint,
+    @startResourceSurrogateId bigint,
+    @endResourceSurrogateId bigint,
+    @batchSize int
+AS
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    DECLARE @Sql NVARCHAR(MAX);
+    DECLARE @ParmDefinition NVARCHAR(512);
+
+    SET @sql = N'DELETE TOP(@BatchSizeParam) FROM ' + @tableName	+ N' WHERE ResourceTypeId = @ResourceTypeIdParam AND ResourceSurrogateId >= @StartResourceSurrogateIdParam AND ResourceSurrogateId < @EndResourceSurrogateIdParam'
+    SET @parmDefinition = N'@BatchSizeParam int, @ResourceTypeIdParam smallint, @StartResourceSurrogateIdParam bigint, @EndResourceSurrogateIdParam bigint'; 
+
+	EXECUTE sp_executesql @sql, @parmDefinition,
+                          @BatchSizeParam = @batchSize,
+                          @ResourceTypeIdParam = @resourceTypeId,
+                          @StartResourceSurrogateIdParam = @startResourceSurrogateId,
+                          @EndResourceSurrogateIdParam = @endResourceSurrogateId
+
+    COMMIT TRANSACTION
+
+    return @@rowcount
+GO
+
+/*************************************************************
+    Stored procedures for disable indexes
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     DisableIndexes
+--
+-- DESCRIPTION
+--     Stored procedures for disable indexes
+--
+-- PARAMETERS
+--     @indexes 
+--         * indexes table
+CREATE OR ALTER PROCEDURE [dbo].[DisableIndexes]
+    @indexes dbo.IndexTableType_1 READONLY
+AS
+    SET NOCOUNT ON
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    declare commands cursor for
+    SELECT N'ALTER INDEX [' + indexes.IndexName + '] ON ' + indexes.TableName + ' Disable;'
```

Comment:
>[' + indexes.IndexName + '] [](http://example.com/codeflow?start=25&length=27)

Dynamic SQL based on user-provided input is risky.  Any alternatives?  Separately, consider using QUOTENAME() instead of hard-coding the brackets (what if indexes.IndexName contains ']'?).

Reply:
Thanks QUOTENAME works here. And for dynamic index name here, it looks like can not verify as tablename...

Since it is only used by our code and indexes names are predefined in code, I think it should be safe for no direct customer use.

Reply:
Thanks, now we check index and table name before use to prevent inject attack.